### DISCUSSION TITLE
Bounce handling is not working | Bounce messages are not forwarded to END POINT/ROUTING ADDRESS

### DISCUSSION THREAD

**Discussion by XenyFord** [postalserver/postal#1567] [Help with using Postal] [answered]
Dear Respected Team,

I don't know whether it's a bug or lack of a feature or any wrong configuration made on my end, but ultimately bounce messages are not forwarded to the Routing/ENDPOINT email ADDRESS, but they are perfectly identified and listed under 'Outgoing Messages' in the postal dashboard.

After everything done by setting-up **_email server_**, **_domain_**, **_user credentials_**, **_Routing_** and **_ENDPOINT email address_**, I made a test sending an email to an invalid email address(**hardbounce@bounce-testing.postmarkapp.com**) using my 'Credentials' and 'FROM address(**no-reply@mywebsite.com**)' on **_https://smtper.net_** , and got bounce message shown in the dashboard under '**_Outgoing Messages_**' marked as **HARD FAIL**, but still the bounce messages are not forwarded to ROUTING/ENDPOINT EMAIL ADDRESS.

## To Reproduce

**Below are the screenshot of the entire process I went through for your reference..**

**Step 1:** Created a mail server named(my email server), domain and credentials as shown in the pics below;
- a. https://prnt.sc/1qjbxkp
- b. https://prnt.sc/1qjc24h
- c. https://prnt.sc/1qjc7az

**Step 2:** I have created ROUTING & ENDPOINT ADDRESSES as shown in the pic below;
- d. https://prnt.sc/1qjhmjr

**Step 3:** made a test sending an email to an invalid address(hardbounce@bounce-testing.postmarkapp.com) as shown in the below pic;
- e. https://prnt.sc/1qjcp7p

**Step 4:** I had the bounce back message displayed under 'Outgoing Messages' in the postal dashboard as shown in the pic below;
- f. https://prnt.sc/1qjcu8r

**Step 5:** No bounce messages are received/forwarded to the email address(no-replay@mywebsite.com) given in the postal settings;
- g. https://prnt.sc/1qjcyz6

## Expected behavior

I think, as usual, It must be allowed us to have bounce messages forwarded to the given Routing/ENDPOINT email address(no-reply@mywebsite.com)... Or It's better allow us to have a custom bounce email address instead of auto generated email(kming@psrp.mywebsite.com) by postal.

## Screenshots

1. https://prnt.sc/1qjbxkp
2. https://prnt.sc/1qjc24h
3. https://prnt.sc/1qjc7az
4. https://prnt.sc/1qjhmjr
5. https://prnt.sc/1qjcp7p
6. https://prnt.sc/1qjcu8r
7. https://prnt.sc/1qjcyz6

## Environment details 

 - OS: [Ubuntu 21.04]
 - Browser [chrome, safari]
 - Version [N/A]
 - Type [desktop, mobile etc...]

## Additional information/context

Please let me have the right solution for this ASAP. Also, if required, I will provide you with all the details like live server login and other credential, etc. to allow you have access to our server to solve this issue, If possible. 

Thank you in advance!

**Comment by adamcooke** [comment, marked as answer]
This is not a bug, you just don't seem to understand what a bounce is. I'm converting this to a discussion as this is not a bug in Postal.

If you have a hard fail, it won't bounce, the SMTP server will just reject it straight away when we connect.

**Reply by XenyFord** [reply]
Hi Adam,

YES! Of course, even if it's not bounce(Hard FAIL) — But, when the server was unable to deliver the message to a specific email address which is an invalid one(example@example.com), it must be returned to us on a specific email address just like how we receive commonly with any other email systems right?

I haven't get any non-delivered emails forwarded to my routing/ENDPOINT addresses since I tested sending messages to many invalid addresses said to be BOUNCE addresses as mentioned in the links below;

List of bounce address for your reference:

https://gist.github.com/cjaoude/fd9910626629b53c4d25
https://postmarkapp.com/support/article/1239-how-to-test-bounces

If it is not a bug/missed feature, please be very kind to let me know on how to have the non-delivered message forwarded to the specific email address given under ROUTING/ENDPOINT, instead of just showing it up in the dashboard as SOFT FAIL/HARD FAIL, please?

Hope you understand.

Thank you in advance!

**Reply by XenyFord** [reply]
> 
> 
> This is not a bug, you just don't seem to understand what a bounce is. I'm converting this to a discussion as this is not a bug in Postal.
> 
> If you have a hard fail, it won't bounce, the SMTP server will just reject it straight away when we connect.

**Or, 
Can you please be very kind to let me know the exact setting that will make me to have the UN-delivered messages forwarded on a specific email address of mine?**

**Reply by XenyFord** [reply]
> 
> 
> This is not a bug, you just don't seem to understand what a bounce is. I'm converting this to a discussion as this is not a bug in Postal.
> 
> If you have a hard fail, it won't bounce, the SMTP server will just reject it straight away when we connect.
=====================>>

**Hats Off..! ...Hats Off..!! ....Hats Off..!!!**

Dear Mr. Adam,

It's a real pleasure for me writing to you here.

I followed your great video guide on the postal official website and similarly I installed postal on Ubuntu 21.04 as you did, it was really much wonderful and more useful/helpful for every beginner. **_Appreciated sir! I'm sure the [video](https://youtu.be/d1Lzw_Q_fJQ) made by you is really amazing and it's very easy to understand the complete steps of postal installation in a very wonderful way._**

As I said above, I followed your great video guide and did everything as you did there. But, my humble request to you to let me know the exact settings / commands to enable TLS for sending emails over a secure connection via TLS/SSL. — **_This will be very much useful not only for me, but also for many others who are seeking for a proper guidance on how to enable TLS/SSL for their postal server._**

Also, I found 3 different files(.crt, .key and .jason) in `/opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/mydomain.com/`, and I copied two of them(.crt and .key) into ` /opt/postal/config/ `, and renamed as `smtp.crt` & `smtp.key` and then mentioned the same into the` postal.yml `file as shown below;

`smtp_server:`
  `port: 25`
  `tls_enabled: true`
  `tls_certificate_path: config/smtp.cert`
  `tls_private_key_path: config/smtp.key`
  `tls_ciphers:`
  `ssl_version: SSLv23`
  `proxy_protocol: false`
  `log_connect: true`
  `strip_received_headers: false`
  `max_message_size: 14`
  
**But, when I made a TEST sending an email via secured connection using TLS, I got few errors as shown in the image attached below;**

![error-new](https://user-images.githubusercontent.com/89290310/132100295-28704be9-35e2-444b-bad9-751338fcda59.jpg)

**Below are the ` smtp.crt ` & ` smtp.key ` files copied from ` caddy-data/../../../mydomain.com `**

![Copied-into-smtp-crt](https://user-images.githubusercontent.com/89290310/132100304-c5c2c3ba-de26-4de9-8c18-d644e87401ed.jpg)
---
![Copied-into-smtp-key](https://user-images.githubusercontent.com/89290310/132100311-3199fb0b-5f82-445d-87a5-140c28c4ba30.jpg)

 
Please be so kind to help me to get this solved ASAP. A heartfelt thanks in advance!

Kind Regards,

**Reply by willpower232** [reply]
do you need to put a `/` before the `config/smtp.whatever`?

**Reply by XenyFord** [reply]
Mr. Will,
It gives a different ERROR as shown in the image below When I replace ` config/smtp.crt ` & ` config/smtp.key ` with ` /config/smtp.crt ` & ` /config/smtp.key `.
![ERROR-NEW](https://user-images.githubusercontent.com/89290310/132251764-bdb8e24c-a1d9-4356-bd14-7a5b0ae6e0ef.jpg)


Please guide me how to solve this. Thank you in advance!

**Reply by willpower232** [reply]
it means that the key file is invalid somehow, there is usually a comment at the top of the file `-----BEGIN RSA PRIVATE KEY-----` or something similar to tell you what kind of file it is

**Reply by XenyFord** [reply]
Mr. Will -

I found only three files(` xyz.crt ` , ` xyz.key ` and ` xyz.jason `) in ` /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/mydomain.com/  ` as shown in the image above, and copied two of them into config folder and renamed to ` smtp.crt ` & ` smtp.key ` and added the same path to postal.yml.

Also, The config folder has three different files except duplicated yml i.e ` postal.yml.save ` file and other copied files - ` smtp.crt ` and ` smtp.key ` as shown in the image.

![SCREEN-Config](https://user-images.githubusercontent.com/89290310/132359473-a98753cc-109f-449d-beab-c9f91f7d6ce1.jpg)

So, please guide me further. Thank you in advance!

**Reply by willpower232** [reply]
If you use the `head` command, i.e. `head smtp.crt` you should be able to see what it really is.

**Reply by XenyFord** [reply]
Mr. Will,

The output is as shown in the image below when I run ` head /opt/postal/config/smtp.crt `

![smtp-cert](https://user-images.githubusercontent.com/89290310/132480334-f288a2b9-db70-4556-b713-7af5b622f8c1.jpg)

If there is any problem with the current certificate, please let me know how to remove the Caddy from the server and use a different SSL / a purchased one instead of it....Please help me.

Awaiting reply. Thank you in advance!

**Reply by XenyFord** [reply]
Mr. Will,
Or, please let me know on how to remove the current SSL(Caddy) and install a new the NEW one? because it seem to be a lot of confusing stuff involved. Thank you in advance!

**Reply by willpower232** [reply]
The key file is apparently the problem one, what is the head of that?

**Reply by XenyFord** [reply]
Mr. Will,
The head of the ` smtp.key ` is as below..

![ECKey](https://user-images.githubusercontent.com/89290310/132638203-0fb184fb-7941-4f8e-a797-9df53f8d98c7.jpg)


Please guide me in this case. Thank you in advance!

**Reply by willpower232** [reply]
Unfortunately Postal can't use that kind of key (#1248), I don't know enough about Caddy to know if you can create and use an RSA-based certificate instead.

**Reply by XenyFord** [reply]
Mr. Will,
No problem I simply switch it back to RSA from EC keys. It's working now. But, few other problems as shown in the images are is still existed..

![ERROR-NEw1](https://user-images.githubusercontent.com/89290310/133075489-89ec9525-7e45-4b34-8b93-0e14b316f832.jpg)
------------------
![ERRO-new2](https://user-images.githubusercontent.com/89290310/133075493-abc4df1b-1b84-4a26-8a9c-32d20eaacde8.jpg)

Could you please let me know on how to solve this? Thank you in advance!
